task/backend/scheduler.go

```diff
@@ -779,36 +786,37 @@ func (r *runner) executeAndWait(ctx context.Context, qr QueuedRun, runLogger *za
 		runLogger.Info("Failed to wait for execution result", zap.Error(err))
 
 		// TODO(mr): retry?
-		r.fail(qr, runLogger, "Waiting for execution result", err)
+		r.fail(qr, runLogger, "Waiting for execution result", *influxdb.ErrRunExecutionError(err))
 		return
 	}
 	if err := rr.Err(); err != nil {
 		runLogger.Info("Run failed to execute", zap.Error(err))
 		errMsg = "Run failed to execute, " + errMsg
 
 		// TODO(mr): retry?
-		r.fail(qr, runLogger, "Run failed to execute", err)
+		r.fail(qr, runLogger, "Run failed to execute", *influxdb.ErrRunExecutionError(err))
 		return
 	}
 
 	stats := rr.Statistics()
 
 	b, err := json.Marshal(stats)
 	if err == nil {
+		err = influxdb.ErrJsonMarshalError(err)
 		// authctx can be updated mid process
 		r.ts.nextDueMu.RLock()
 		authCtx := r.ts.authCtx
 		r.ts.nextDueMu.RUnlock()
 		r.taskControlService.AddRunLog(authCtx, r.task.ID, qr.RunID, time.Now(), string(b))
 	}
-	r.updateRunState(qr, RunSuccess, runLogger)
+	r.updateRunState(qr, RunSuccess, runLogger, err.(*influxdb.Error))
```

Comment:
If an error is returned from `json.Marshal`, this will panic as `err` will not be an instance of `*influxdb.Error`

An alternative approach

```go
var influxErr *influxdb.Error
b, err := json.Marshal(stats)
if err != nil {
    influxErr = influxdb.ErrJsonMarshalError(err)
} else {
	r.ts.nextDueMu.RLock()
	authCtx := r.ts.authCtx
	r.ts.nextDueMu.RUnlock()
	r.taskControlService.AddRunLog(authCtx, r.task.ID, qr.RunID, time.Now(), string(b))
}
r.updateRunState(qr, RunSuccess, runLogger, influxErr)
```

Reply:
ohh nice catch, I mistook the `==` for a `!=`